README.md
# scripts

Useful scripts

 Script Name | Description
 ------ | ------
addLoopback | Powershell to add a loopback interface to solve "the ARP problem"
isIPFree | simple script to see if an IP address is in use on the network
HealthChecks | Loadbalancer / Monitoring health check scripts


isIPFree/CHANGELOG
Version 0.01

Initial release.  Support for IPv4 and IPv6 addresses


isIPFree.sh
#!/bin/bash

# Simple script to check if an IP is available or not
# Version 0.01 - Jon Morby 5th November 2022
#
# exit with 1 if an IP is found, 0 if success

# Variables
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin

DATE=`date +%b\ %d\ %H:%M:%S` # syslog format
APP=$(basename ${0})
LOG=${APP}.log
PING="ping"
TIMEOUT=5
DEBUG=0
IPADDR=${1}

#### functions

function ip_found {
	msg="FAIL: IP ${IPADDR} seems to be in use, try another"
	echo "${DATE} ${APP} ${msg}" >> /var/log/${LOG}
	echo -e "\r\n\r\n${msg}"
	exit 1;
}

##### 

if [ -z ${1} ];
then
	echo -e "Usage:\r\n${0} IP.Add.re.ss\r\n\r\nExample:\r\n\r\n\t${0} 192.168.1.10\r\n"
	exit 1;
fi

# is this an IPv6 address?

if [[ ${1} == *":"* ]];
then
	ARGS="-6"
	PING="ping6"
fi

echo "Checking to see if IP address $1 is currently in use on the network"

if [ ${DEBUG} -eq "1" ]; then
	echo "DEBUG: ping"
fi

${PING} -q -c 4 ${1} &> /dev/null

if [ ${?} -eq "0" ];
then
	ip_found
	exit 1;
fi


RESULT=$( timeout ${TIMEOUT} nmap ${ARGS} ${1} | grep -c "Host is up" )

if [ ${DEBUG} -eq "1" ]; then
	echo "DEBUG: nmap"
fi

if [ ${RESULT} -eq "1" ];
then
	ip_found
	exit 1;
fi

if [ ${DEBUG} -eq "1" ]; then
	echo "DEBUG: ip neigh show"
fi

RESULT=$(ip ${ARGS} neigh show ${1} | grep -c ${IPADDR} )

if [ ${RESULT} -eq "1" ];
then
	msg="SUCCESS: IP ${IPADDR} is Probably available"
        echo "${DATE} ${APP} ${msg}" >> /var/log/${LOG}
	echo -e "\r\n\r\n${msg}"
	exit 0;
fi

#No Joy, Failure, IP is in use on the network
ip_found
exit 1


Addloopback/AddLoopback.ps1
## AddLoopback.ps1 by Jon Morby
## (C) Copyright 2022
## All Rights Reserved
##
## Version Info
## 0.01 - Initial version (beta)
## 0.02 - IPv6 support enhancements

param(
    [Parameter(Mandatory=$false)][String] $ip4,
    [Parameter(Mandatory=$false)][String] $ip6,
    [Parameter(Mandatory=$false)][String] $ip,
    [Parameter(Mandatory=$false)][String] $ethernet = 'Ethernet0',
    [Parameter(Mandatory=$false)][switch] $print,
    [Parameter(Mandatory=$false)][switch] $help,
    [Parameter(Mandatory=$false)][switch] $cleanup,
    [Parameter(Mandatory=$false)][switch] $v = $false,
    [Parameter(Mandatory=$false)][switch] $vv = $false # Quit before actually installing anything
)

#Requires -RunAsAdministrator
$Version = '0.02 beta'
$scriptName = $MyInvocation.MyCommand

$primary_interface = $ethernet

function showUsage {
    showBanner

    Write-Host "Usage: $scriptName parameters
    
    -ip <IP>        ip address for loopback adapter
    -print          enable printer support
    -ethernet       Name of Ethernet Interface - default is "Ethernet0"
        
    Examples: 
        $scriptName -ip4 192.168.1.1 -print
        $scriptName -ip4 192.168.1.1
        $scriptName -ip6 2001:0db8:85a3:0000:0000:8a2e:0370:7334

    
    "

    Exit 1
}

function showBanner  {
    Write-Host "
Powershell tool to add loopback adapter to a Windows Server. 
Version $Version   
    "
}


# The name for the loopback adapter interface that will be created.
$loopback_name = 'Loopback'

# The name for the servers main network interface. This will be updated to allow weak host send/recieve which is most likely required for the traffic to work for the loopback interface.
# $primary_interface = 'Ethernet0'

if ($help) {
    showUsage
    Exit 0
}

if ($ip) {
    # this really should just check if it is v4 or v6 and act accordingly, but for now bounce the request and ask for it to be re-formed
    Write-Host "Incorrect option -ip please say -ip4 or -ip6"
    showUsage
    Exit 0
}

if ($cleanup) {
   # Remove-NetIPAddress
    Get-NetIPAddress -InterfaceAlias $loopback_name | Remove-NetIPAddress -confirm $false
    Remove-LoopbackAdapter ($loopback_name)
    Exit 0
}

if (!$ip4 -and !$ip6) {
    showBanner
    
    $ip = $(Read-Host "Enter IP address for Loopback Interface")

    if ($ip -like "*:*") {
        $ip6 = $ip
    } else {
        $ip4 = $ip
    }
}

if ($ip4) {

# The IPv4 address that you would like to assign to the loopback interface along with the prefix length (eg. if the IP is routed to the server usually you would set the prefix length to 32).
    $loopback_ipv4 = $ip4
    $loopback_ipv4_length = '32'
}

if ($ip6) {

# The IPv6 address that you would like to assign to the loopback interface along with the prefix length (eg. if the IP is routed to the server usually you would set the prefix length to 128). If you are not adding an IPv6 address do not set these variables.
    $loopback_ipv6 = $ip6
    $loopback_ipv6_length = '128'
}

if ($v -or $vv) {
    Write-Host "DEBUG:
    IP Entered: $ip
    IPv4: $loopback_ipv4 / $loopback_ipv4_length
    IPv6: $loopback_ipv6 / $loopback_ipv6_length
    "
}

if ($vv) {
    # Don't actually install the modules or make changes - debug mode
    Exit 5
}

Install-Module -Name LoopbackAdapter -MinimumVersion 1.2.0.0 -Force
Import-Module -Name LoopbackAdapter

New-LoopbackAdapter -Name $loopback_name -Force

$interface_loopback = Get-NetAdapter -Name $loopback_name
$interface_main = Get-NetAdapter -Name $primary_interface

Set-NetIPInterface -InterfaceIndex $interface_loopback.ifIndex -InterfaceMetric "254" -WeakHostReceive Enabled -WeakHostSend Enabled -DHCP Disabled
Set-NetIPInterface -InterfaceIndex $interface_main.ifIndex -WeakHostReceive Enabled 
Set-NetIPAddress -InterfaceIndex $interface_loopback.ifIndex -SkipAsSource $True
Get-NetAdapter $loopback_name | Set-DNSClient -RegisterThisConnectionsAddress $False

if ($ip4) {
    # Set the IPv4 address
    New-NetIPAddress -InterfaceAlias $loopback_name -IPAddress $loopback_ipv4 -PrefixLength $loopback_ipv4_length -AddressFamily ipv4
}

if ($ip6) {
    # Set the IPv6 address - Uncomment this if required
    New-NetIPAddress -InterfaceAlias $loopback_name -IPAddress $loopback_ipv6 -PrefixLength $loopback_ipv6_length -AddressFamily ipv6
}

#Turn off all bindings on the loopback interface 
Disable-NetAdapterBinding -Name $loopback_name 

if ($ip4) {
    # add ipv4 support back in
    Enable-NetAdapterBinding -Name $loopback_name -ComponentID ms_tcpip
}
if ($ip6) {
    # add ipv6 support back in
    Enable-NetAdapterBinding -Name $loopback_name -ComponentID ms_tcpip6
}


if ($print) {
    # add in support for file and printer sharing as we're in a print environment

    Enable-NetAdapterBinding -Name $loopback_name -ComponentID ms_msclient
    Enable-NetAdapterBinding -Name $loopback_name -ComponentID ms_server
}




HealthChecks/tcp_half_open.sh
#!/bin/bash
# Version 0.1 - Unknown
# Version 0.2 - Basic half open check with some logging / debug options based on one of the WHI health checks
#               Jon Morby - solutions@loadbalancer.org
# Version 0.3 - Neil Stone <support@loadbalancer.org.> - Improvements on real server port handling
# Version 0.4 - Jon Morby - <solutions@loadbalancer.org> - Tidy up date formatting

PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin

### Variables
DEBUG=1
DATE=`date +%b\ %d\ %H:%M:%S` # syslog format
APP=$(basename ${0})
LOG=${APP}.log
RIP="${3}"    # $3 represents the real server IP address as passed by the load balancer
TIMEOUT="3"   # Timeout for checking each port

if [ "${4}" == "0" ] || [ -z "${4}" ]; then
    HALF_OPEN_RPT="${2}"
else
    HALF_OPEN_RPT="${4}"
fi

### Shouldn't need to edit below here

for port in ${HALF_OPEN_RPT}; do
        timeout ${TIMEOUT} nmap -sS -p ${port} ${RIP} 2>&1 | grep -q 'open'
        ec=${?}
        if [ ${ec} -ne "0" ]; then
            if [ ${DEBUG} -eq "1" ]; then
                echo "${DATE} ${APP} Fail: ${RIP}:${port} (${ec})" >> /tmp/${LOG}
            fi
                exit ${ec}
        fi
done

if [ ${DEBUG} -eq "1" ]; then
        echo "${DATE} ${APP} Success: ${RIP}:${port} (${ec})" >> /tmp/${LOG}
fi

exit 0


addLoopback/addLoopback.ps1
## AddLoopback.ps1 by Jon Morby
## (C) Copyright 2022
## All Rights Reserved
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#     Unless required by applicable law or agreed to in writing, software
#     distributed under the License is distributed on an "AS IS" BASIS,
#     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#     See the License for the specific language governing permissions and
#     limitations under the License.
#
##
## Version Info
## 0.01 - Initial version (beta)
## 0.02 - IPv6 support enhancements
## 0.03 - hidden cleanup options

#Requires -RunAsAdministrator

param(
    [Parameter(Mandatory=$false)][String] $ip4,
    [Parameter(Mandatory=$false)][String] $ip6,
    [Parameter(Mandatory=$false)][String] $ip,
    [Parameter(Mandatory=$false)][String] $ethernet = 'Ethernet0',
    [Parameter(Mandatory=$false)][switch] $print,
    [Parameter(Mandatory=$false)][switch] $help,
    [Parameter(Mandatory=$false)][switch] $cleanup,
    [Parameter(Mandatory=$false)][switch] $v = $false,
    [Parameter(Mandatory=$false)][switch] $vv = $false # Quit before actually installing anything
)

$Version = '0.03 beta'
$scriptName = $MyInvocation.MyCommand

$primary_interface = $ethernet

function showUsage {
    showBanner

    Write-Host "Usage: $scriptName parameters
    
    -ip4 <IP>       IPv4 address for loopback adapter
    -ip6 <IP>       IPv6 format address for loopback adapter
    -print          enable printer support
    -ethernet       Name of Ethernet Interface - default is "Ethernet0"
        
    Examples: 
        $scriptName -ip4 192.168.1.1 -print
        $scriptName -ip4 192.168.1.1
        $scriptName -ip6 2001:0db8:85a3:0000:0000:8a2e:0370:7334

    
    "

    Exit 1
}

function showBanner  {
    Write-Host "
Powershell tool to add loopback adapter to a Windows Server. 
Version $Version   
    "
}


# The name for the loopback adapter interface that will be created.
$loopback_name = 'Loopback'

# The name for the servers main network interface. This will be updated to allow weak host send/recieve which is most likely required for the traffic to work for the loopback interface.
# $primary_interface = 'Ethernet0'

if ($help) {
    showUsage
    Exit 0
}

if ($ip) {
    # this really should just check if it is v4 or v6 and act accordingly, but for now bounce the request and ask for it to be re-formed
    Write-Host "Incorrect option -ip please say -ip4 or -ip6"
    showUsage
    Exit 0
}

if ($cleanup) {
   # Remove-NetIPAddress
    Get-NetIPAddress -InterfaceAlias $loopback_name | Remove-NetIPAddress -confirm $false
    Remove-LoopbackAdapter ($loopback_name)
    Exit 0
}

if (!$ip4 -and !$ip6) {
    showBanner
    
    $ip = $(Read-Host "Enter IP address for Loopback Interface")

    if ($ip -like "*:*") {
        $ip6 = $ip
    } else {
        $ip4 = $ip
    }
}

if ($ip4) {

# The IPv4 address that you would like to assign to the loopback interface along with the prefix length (eg. if the IP is routed to the server usually you would set the prefix length to 32).
    $loopback_ipv4 = $ip4
    $loopback_ipv4_length = '32'
}

if ($ip6) {

# The IPv6 address that you would like to assign to the loopback interface along with the prefix length (eg. if the IP is routed to the server usually you would set the prefix length to 128). If you are not adding an IPv6 address do not set these variables.
    $loopback_ipv6 = $ip6
    $loopback_ipv6_length = '128'
}

if ($v -or $vv) {
    Write-Host "DEBUG:
    IP Entered: $ip
    IPv4: $loopback_ipv4 / $loopback_ipv4_length
    IPv6: $loopback_ipv6 / $loopback_ipv6_length
    "
}

if ($vv) {
    # Don't actually install the modules or make changes - debug mode
    Exit 5
}

Install-Module -Name LoopbackAdapter -MinimumVersion 1.2.0.0 -Force
Import-Module -Name LoopbackAdapter

New-LoopbackAdapter -Name $loopback_name -Force

$interface_loopback = Get-NetAdapter -Name $loopback_name
$interface_main = Get-NetAdapter -Name $primary_interface

Set-NetIPInterface -InterfaceIndex $interface_loopback.ifIndex -InterfaceMetric "254" -WeakHostReceive Enabled -WeakHostSend Enabled -DHCP Disabled
Set-NetIPInterface -InterfaceIndex $interface_main.ifIndex -WeakHostReceive Enabled 
Set-NetIPAddress -InterfaceIndex $interface_loopback.ifIndex -SkipAsSource $True
Get-NetAdapter $loopback_name | Set-DNSClient -RegisterThisConnectionsAddress $False

if ($ip4) {
    # Set the IPv4 address
    New-NetIPAddress -InterfaceAlias $loopback_name -IPAddress $loopback_ipv4 -PrefixLength $loopback_ipv4_length -AddressFamily ipv4
}

if ($ip6) {
    # Set the IPv6 address - Uncomment this if required
    New-NetIPAddress -InterfaceAlias $loopback_name -IPAddress $loopback_ipv6 -PrefixLength $loopback_ipv6_length -AddressFamily ipv6
}

#Turn off all bindings on the loopback interface 
Disable-NetAdapterBinding -Name $loopback_name 

if ($ip4) {
    # add ipv4 support back in
    Enable-NetAdapterBinding -Name $loopback_name -ComponentID ms_tcpip
}
if ($ip6) {
    # add ipv6 support back in
    Enable-NetAdapterBinding -Name $loopback_name -ComponentID ms_tcpip6
}


if ($print) {
    # add in support for file and printer sharing as we're in a print environment

    Enable-NetAdapterBinding -Name $loopback_name -ComponentID ms_msclient
    Enable-NetAdapterBinding -Name $loopback_name -ComponentID ms_server
}


